### ISSUE TITLE
Flang: change behaviour for unsupported targets

### ISSUE THREAD

**Issue by rjzak** [llvm/llvm-project#55015]
Currently, trying to use an unsupported target asks the user to submit a bug report (see #54832).
Instead, `flang` should exit gracefully after indicating that the requested target isn't supported, and perhaps display a list of supported targets.

**Comment by banach-space** [comment]
Thanks for creating this ticket!

In #54832, you were hitting an LLVM fatal error from [here](https://github.com/llvm/llvm-project/blob/3d0e0e1027203fe5e89104ad81ee7bb53e525f95/flang/lib/Optimizer/CodeGen/Target.cpp#L291):
```cpp
llvm::report_fatal_error("target not implemented");
```
That's not part of the driver, so let me re-label this (solving this might require changing other parts of Flang).

I agree with you that for unsupported targets, Flang should warn users rather than crash. I see two possible approaches to improve this (perhaps others will have other suggestions):

1. Add the default case in [fir::CodeGenSpecifics::get](https://github.com/llvm/llvm-project/blob/3d0e0e1027203fe5e89104ad81ee7bb53e525f95/flang/lib/Optimizer/CodeGen/Target.cpp#L243-L292). 
2.  Add a hook to identify whether the requested triple is supported , e.g. `isTargetSupported(llvm::Triple)` and  use it in `fir::CodeGenSpecifics::get` and somewhere in the driver to exit early with a "nice" diagnostic.

Option 1. is the least intrusive approach, but it will make Flang generate code for platforms that it perhaps shouldn't. I'm not 100% whether that would be safe/desirable. Option 2. would require a bit more code, but won't change the current behavior (i.e. for "unsupported" triples Flang will just refuse to generate code). 

-Andrzej

**Comment by banach-space** [comment]
FYI: https://reviews.llvm.org/D126537

**Comment by xgupta** [comment]
> @xgupta , thanks for revisiting this 🏅 !

No problem, I guess I have joined Flang/Fortran team :-)
